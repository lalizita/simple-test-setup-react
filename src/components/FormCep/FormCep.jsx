import React, { useState } from 'react';
import { Box, Flex } from 'rebass'
import {
  Label,
  Input,
  Select,
  Textarea,
  Radio,
  Checkbox,
} from '@rebass/forms';
import axios from 'axios'

const FormCep = () => {
  const initialAddress = {
    logradouro: '',
    bairro: ''
  }
  const [ cep, setCep ] = useState('')
  const [ address, setAddress ] = useState(initialAddress)

  const fetchCep = async (cep) => {
    const cepResult = await fetch(`https://viacep.com.br/ws/${cep}/json/`)
    const { logradouro, bairro } = await cepResult.json()
    setAddress({ logradouro, bairro })
  }

  const handleCepField = value => {
    if(value.length === 8) fetchCep(value)
    setCep(value)
  }

  return (
    <Box as='form' onSubmit={(e) => e.preventDefault()} py={3} width='500px'>
      <Flex mx={-2} mb={3}>
        <Box width={1 / 2} px={2}>
        <Label htmlFor='cep'>Cep</Label>
          <Input
            id='cep'
            name='cep'
            placeholder="CEP"
            onChange={({ target : { value } }) => handleCepField(value)}
          />
        </Box>
        <Box width={1 / 2} px={2}>
        <Label htmlFor='bairro'>Bairro</Label>
          <Input
            id='bairro'
            name='bairro'
            placeholder="bairro"
            value={address.bairro}
            onChange={({ target : { value } }) => setAddress({...address, bairro: value})}
          />
        </Box>
        <Box width={1 / 2} px={2}>
        <Label htmlFor='logradouro'>Logradouro</Label>
          <Input
            id='logradouro'
            name='logradouro'
            placeholder="logradouro"
            value={address.logradouro}
            onChange={({ target : { value } }) => setAddress({...address, logradouro: value})}
          />
        </Box>
      </Flex>
    </Box>
  );
};

export default FormCep;
