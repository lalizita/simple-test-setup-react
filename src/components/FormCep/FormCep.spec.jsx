import React from 'react'
import { render, fireEvent, getByText } from '@testing-library/react'
import '@testing-library/jest-dom/extend-expect'
import { act } from 'react-dom/test-utils';
import FormCep from './FormCep'

const addressResponse = {
  cep: '03047000',
  logradouro: 'Rua x',
  bairro: 'Novo rioe',
}

global.fetch = jest.fn().mockImplementation(() =>
  Promise.resolve({
    json: () => Promise.resolve(addressResponse)
  })
);

it('Should render cep form', async () => {
  const { debug, getByPlaceholderText, container } = render(<FormCep />)
  const cepInput = getByPlaceholderText('CEP').closest('input')
  fireEvent.change(cepInput, { target: { value: '03047000' } })

  await act(() => global.fetch)

  expect(global.fetch).toHaveBeenCalledTimes(1)
  expect(container).toMatchSnapshot()
})