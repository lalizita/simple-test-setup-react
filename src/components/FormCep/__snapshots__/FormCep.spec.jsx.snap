// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should render cep form 1`] = `
<div>
  <form
    class="css-1ldrcmr"
  >
    <div
      class="css-1viozcy"
    >
      <div
        class="css-1e8zq2i"
      >
        <label
          class="css-knfhfe"
          for="cep"
        >
          Cep
        </label>
        <input
          class="css-4y926w"
          id="cep"
          name="cep"
          placeholder="CEP"
          type="text"
        />
      </div>
      <div
        class="css-1e8zq2i"
      >
        <label
          class="css-knfhfe"
          for="bairro"
        >
          Bairro
        </label>
        <input
          class="css-4y926w"
          id="bairro"
          name="bairro"
          placeholder="bairro"
          type="text"
          value="Novo rio"
        />
      </div>
      <div
        class="css-1e8zq2i"
      >
        <label
          class="css-knfhfe"
          for="logradouro"
        >
          Logradouro
        </label>
        <input
          class="css-4y926w"
          id="logradouro"
          name="logradouro"
          placeholder="logradouro"
          type="text"
          value="Rua x"
        />
      </div>
    </div>
  </form>
</div>
`;
